src/storage/buffer_pool_manager/buffer_pool.rs

```diff
@@ -0,0 +1,432 @@
+            unsafe {
+                // Safety: We unlock the inner lock first.
+                self.frame_read_lock_raw.unlock_shared();
+                self.page_table_item_read_lock_raw.unlock_shared();
+            }
+        }
+    }
+
+    pub type BufferPoolKey = (TableId, PageId);
+
+    pub trait BufferPool: Send + Sync + 'static {
+        fn get_page(&self, id: BufferPoolKey) -> PageReadLock;
+        /// # Safety
+        /// The buffer pool must contain `id`.
+        /// `offset % page_size == 0`.
+        /// `offset + data.len() <= page_size`.
+        fn mutate_page(&self, id: BufferPoolKey, offset: usize, data: Vec<u8>);
+    }
+}
+
+#[derive(Debug)]
+#[repr(u8)]
+/* private */
+enum PageTableItemMetadataFlagsMasks {
+    IsDirty = 0b00000001,
+}
+
+#[derive(Default, Debug)]
+#[repr(transparent)]
+/* private */
+struct PageTableItemMetadataFlags(u8);
+
+macro_rules! create_page_table_item_metadata_flags_get_set {
+    ($mask: expr, $get_name: ident, $set_name: ident) => {
+        pub const fn $get_name(&self) -> bool {
+            (self.0 & ($mask as u8)) != 0
+        }
+
+        pub fn $set_name(&mut self, value: bool) {
+            self.0 = if value {
+                self.0 | ($mask as u8)
+            } else {
+                self.0 | (!($mask as u8))
+            };
+        }
+    };
+}
+
+impl PageTableItemMetadataFlags {
+    create_page_table_item_metadata_flags_get_set!(
+        PageTableItemMetadataFlagsMasks::IsDirty,
+        get_is_dirty,
+        set_is_dirty
+    );
+}
+
+#[derive(Default, Debug)]
+/* private */
+struct PageTableItemMetadata {
+    access_count: usize,
+    pin_count: usize,
+    flags: PageTableItemMetadataFlags,
+}
+
+#[derive(Debug)]
+/* private */
+enum PageTableItem {
+    Free {
+        next_free_index: Mutex<Option<usize>>,
+    },
+    Occupied {
+        metadata: Mutex<PageTableItemMetadata>,
+        frame_lock: RwLock<()>,
+    },
+}
+
+impl PageTableItem {
+    pub fn new_free(next_free_index: Option<usize>) -> Self {
+        Self::Free {
+            next_free_index: Mutex::new(next_free_index),
+        }
+    }
+
+    pub fn new_occupied() -> Self {
+        Self::Occupied {
+            metadata: Mutex::new(PageTableItemMetadata::default()),
+            frame_lock: RwLock::new(()),
+        }
+    }
+}
+
+#[derive(Debug)]
+/* private */
+struct PageTable {
+    root: Mutex<usize>,
+    table: UnsafeCell<Vec<RwLock<PageTableItem>>>,
+}
+
+impl PageTable {
+    pub fn new(entries_count: usize) -> Self {
+        assert!(entries_count > 0);
+
+        Self {
+            root: Mutex::new(0),
+            // TODO: Look into eager initializing this.
+            table: UnsafeCell::new(Vec::with_capacity(entries_count)),
+        }
+    }
+}
+
+unsafe impl Send for PageTable {}
+unsafe impl Sync for PageTable {}
+
+#[derive(Debug)]
+/* private */
+struct FrameBuffer {
+    pub page_size: usize,
+    buffer: UnsafeCell<Vec<u8>>,
```

Comment:
Did you try to explore a way to use a struct rather than u8? It would make the 1 to 1 correspondence with PageTable more clear